File: day_01/python/part2.py
##############################################################################
# 2018 Advent of Code
# Day: 1
# Part: 2
# Answer: 78724
#
# @author Matt Adorjan
##############################################################################

# Read frequency changes in to a variable
with open('../input.txt', 'r') as f:
    data_in = f.readlines()

# Set initial frequecy as 0
FREQ = 0
FREQ_LIST = [FREQ]
CHANGE_IND = 0

# For each of the frequency changes, add to the current frequency value
while 1:
    FREQ = FREQ + int(data_in[CHANGE_IND].strip())    # Convert from str -> int and strip off line endings (\n)

    if FREQ in FREQ_LIST:
        print("The first frequency to appear twice is: {}".format(FREQ))
        break
    
    FREQ_LIST.append(FREQ)

    CHANGE_IND = CHANGE_IND + 1

    if CHANGE_IND == len(data_in):
        CHANGE_IND = 0

File: day_01/go/part2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"strconv"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func stringInSlice(a string, list []string) bool {
    for _, b := range list {
        if b == a {
            return true
        }
    }
    return false
}

func main() {
	dat, err := ioutil.ReadFile("../input.txt")
	check(err)
	lines := strings.Split(string(dat), "\n")

	var FREQ int64 = 0
	FREQ_LIST := make([]string, 0)
	FREQ_LIST = append(FREQ_LIST, "0")
	var CHANGE_IND int = 0

	for {
		line, err := strconv.ParseInt(lines[CHANGE_IND], 10, 64)
		check(err)
		FREQ = FREQ + line

		if stringInSlice(string(FREQ), FREQ_LIST) {
			fmt.Println("The first frequency to appear twice is: ", FREQ)
			break
		}

		// Convert FREQ back to string to add to slice
		fmt.Println(FREQ)
		FREQ := strconv.FormatInt(FREQ, 16)
		fmt.Println(FREQ)
		FREQ_LIST = append(FREQ_LIST, FREQ)
		CHANGE_IND = CHANGE_IND + 1

		if CHANGE_IND == len(FREQ_LIST) {
			CHANGE_IND = 0
		}
	}
}